README
Ruby's Yahoo Finance Wrapper
============================

A dead simple wrapper for yahoo finance quotes end-point.

Usage:

1 - Getting latest quotes for a set of symbols
----------------------------------------------
Just pass an array of symbols (stock names, indexes, exchange rates) and a list of fields you want:

data = YahooFinance.quotes(["BVSP", "NATU3.SA", "USDJPY=X"], [:ask, :bid, :last_trade_date])
# ask will be a string here:
puts data[0].symbol + " value is: " + data[0].ask 

Or optionally:

data = YahooFinance.quotes(["BVSP", "NATU3.SA", "USDJPY=X"], [:ask, :bid, :last_trade_date], { :raw => false } )
# ask will be a float here:
puts data[0].symbol + " value is: " + data[0].ask 

The full list of fields follows:

     :ask
     :average_daily_volume
     :ask_size
     :bid
     :ask_real_time
     :bid_real_time
     :book_value b4
     :bid_size b6
     :chance_and_percent_change
     :change
     :comission
     :change_real_time
     :after_hours_change_real_time
     :dividend_per_share
     :last_trade_date
     :trade_date
     :earnings_per_share
     :error_indicator 
     :eps_estimate_current_year 
     :eps_estimate_next_year 
     :eps_estimate_next_quarter 
     :float_shares 
     :low 
     :high 
     :low_52_weeks 
     :high_52_weeks 
     :holdings_gain_percent 
     :annualized_gain 
     :holdings_gain 
     :holdings_gain_percent_realtime 
     :holdings_gain_realtime 
     :more_info 
     :order_book 
     :market_capitalization 
     :market_cap_realtime 
     :ebitda 
     :change_From_52_week_low 
     :percent_change_from_52_week_low 
     :last_trade_realtime_withtime 
     :change_percent_realtime 
     :last_trade_size 
     :change_from_52_week_high 
     :percent_change_from_52_week_high 
     :last_trade_with_time 
     :last_trade_price
     :close 
     :high_limit 
     :low_limit 
     :days_range
     :days_range_realtime 
     :moving_average_50_day 
     :moving_average_200_day 
     :change_from_200_day_moving_average 
     :percent_change_from_200_day_moving_average 
     :change_from_50_day_moving_average 
     :percent_change_from_50_day_moving_average 
     :name 
     :notes 
     :open 
     :previous_close 
     :price_paid 
     :change_in_percent 
     :price_per_sales 
     :price_per_book 
     :ex_dividend_date
     :pe_ratio 
     :dividend_pay_date 
     :pe_ratio_realtime 
     :peg_ratio 
     :price_eps_estimate_current_year 
     :price_eps_Estimate_next_year 
     :symbol 
     :shares_owned 
     :short_ratio 
     :last_trade_time 
     :trade_links 
     :ticker_trend 
     :one_year_target_price 
     :volume
     :holdings_value 
     :holdings_value_realtime 
     :weeks_range_52 
     :day_value_change 
     :day_value_change_realtime 
     :stock_exchange 
     :dividend_yield 


2 - Getting historical quotes
-----------------------------

Here you can specify a date range and a symbol, and retrieve historical data for it. 
The last parameter (options) can include, besides the "raw" option, a "period" option.
The period can be specified as :daily, :monthly, :weekly or :dividends_only

data = YahooFinance.historical_quotes("BVSP", Time::now-(24*60*60*10), Time::now) # 10 days worth of data

or

data = YahooFinance.historical_quotes("BVSP", Time::now-(24*60*60*10), Time::now, { :raw => false, :period => :monthly })





Enjoy! :-)

- Herval (hervalfreire@gmail.com)


HISTORY
3.0.0-alpha.1 - November 19, 2017
====================
* Code extracted into modules Quotes, Historical, Splits, Symbols
* Test/Unit moved over to Minitest
* Tests extracted out my module
* Add Byebug for debugging
* Add Yard for documentation
* Change Quotes module to use the V7 Api
* Fix Historical per https://github.com/herval/yahoo-finance/pull/53

2.0.0 - Sep 08, 2017
====================
* Replaced broken CSV API (historical data) with parsing Yahoo Finance HTML using Nokogiri (h/t @zolrag13)
* Added support for fetching companies, sectors and industries (h/t @zolrag13)

1.2.0 - Feb 04, 2017
====================
* Fixed support for { raw: false }
* Symbols that are not found will now not return (instead of returning a struct with an empty :symbol field and a bunch of "N/A"s)
* The returned quotes now will always include the :symbol name 

0.3.0 - Aug 16, 2015
====================
* Refactorings & support for US companies data (via NASDAQ's Apis) - by Eric D. Santos Sosa

0.2.0 - Nov 24, 2014
====================
* Allow fetching a single symbol with the `quote` method
* Multiple fixes to comply w/ Yahoo API changes
* Optional start/end dates (the quotes method now takes an options hash)

0.1.0 - Apr 05, 2011
===============================
* Use ruby CSV
* Refactor

0.0.2 - Aug 05, 2007
===============================
* Type-casting: pass the option :raw => false to the quotes method to get columns converted from string to specific types (mostly float, date, time and datetime)
* Historical quotes: a convenient method to get historical quotes for a given symbol in an interval, getting either daily, weekly or monthly quotes - or dividend yields only

0.0.1 - Aug 04, 2007
===============================
* Initial implementation - no bells, no whistles :)


lib/yahoo-finance/historical.rb
require 'nokogiri'

module YahooFinance
  module Historical
    HISTORICAL_MODES = {
      daily: "1d",
      weekly: "1wk",
      monthly: "1mo"
    }

    def historical_quotes(symbol, options = {})
      options[:period] ||= :daily
      start_date = options[:start_date] || Time.now() - time_from_days(1)
      end_date = options[:end_date] || Time.now()
      params = {
        interval: HISTORICAL_MODES[options[:period]],
        filter: 'history'
      }
      days_per_page = case params[:interval]
                      when HISTORICAL_MODES[:daily]
                        140 # 100 results divided by 5 trading days/week * 7 days/week = number of days per results
                      when HISTORICAL_MODES[:weekly]
                        700 # 7days/week * 100 results
                      when HISTORICAL_MODES[:monthly]
                        2900 # for simplicity assume that all months have 29 days * 100 results (remove duplicates at the end)
                      end
      current_date = start_date.to_time.to_i
      data = []
      while Time.at(current_date) < end_date
        params[:period1] = current_date
        current_end_date = [current_date + time_from_days(days_per_page), end_date.to_time.to_i].min
        params[:period2] = current_end_date
        url = "https://finance.yahoo.com/quote/#{URI.escape(symbol)}/history/?#{params.map{|k, v| "#{k}=#{v}"}.join("&")}"
        data << read_historical(symbol, url)
        current_date = current_end_date + time_from_days(1)
      end
      data.uniq.flatten
    end


    private

    def time_from_days(days)
      24*60*60*days
    end

    def read_historical(symbol, url)
      doc = Nokogiri::HTML(open(url))
      rows = doc.xpath("//table")[0].css('tr')

      return [] if rows.empty?
      cols = rows[0].css('th').to_a
      trade_date_col = cols.index { |c| c.text == 'Date' }
      open_col = cols.index { |c| c.text == 'Open' }
      high_col = cols.index { |c| c.text == 'High' }
      low_col = cols.index { |c| c.text == 'Low' }
      close_col = cols.index { |c| c.text == 'Close*' }
      adjusted_close_col = cols.index { |c| c.text == 'Adj Close**' }
      volume_col = cols.index { |c| c.text == 'Volume' }

      rows.drop(0).inject([]) do |data, row|
        divs = row.css('td')
        if divs[1] && !divs[1].text.include?('Dividend') && !divs[1].text.include?('Stock Split') #Ignore these rows in the table
          data << OpenStruct.new({
            'symbol': symbol,
            'trade_date': Date.parse(divs[trade_date_col]).to_s,
            'open': divs[open_col].text.to_f,
            'high': divs[high_col].text.to_f,
            'low': divs[low_col].text.to_f,
            'close': divs[close_col].text.to_f,
            'adjusted_close': divs[adjusted_close_col].text.to_f,
            'volume': divs[volume_col].text.gsub(/[^\d^\.]/, '').to_f
          })
        end
        data
      end
    end
  end
end
